### PR TITLE
ffmpeg: Re-write the module

### PR DESCRIPTION
**Pull request by yshalsager**
## Changes:

- There's no need to use `ffmpegsave` command anymore. All module commands will work with telegram files, local files, and direct links.
- `ffmpegtrim` has been replaced with `atrim` or `vtrim` command.

## New commands

- `x256` to encode videos using libx265, with the ability to use custom crf using `-b` flag.
- `v2a` to convert a video to audio, with the ability to use custom bitrate using `-b` flag.
- `vscale` to change video quality, with the ability to use custom crf using `-b` flag and encode with x256 using `-c` flag.
- `vth` to extract video thumbnail.
- `vcompress` to compress a video file to a given percentage.
- `minfo` to get media info using ffmpeg.

### CODE REVIEW

File: plugins/ff_mpeg.py [hunk 1]
```diff
@@ -1,156 +1,350 @@
...
+    if not dl_loc or not file_name:
+        return
+    FF_MPEG_DOWN_LOAD_MEDIA_PATH.mkdir(parents=True, exist_ok=True)
+    video_file = f"{FF_MPEG_DOWN_LOAD_MEDIA_PATH}/x256_{file_name}"
+    await message.edit("`Encoding to x256...`")
+    start = datetime.now()
+    try:
+        ffmpeg.input(dl_loc).output(
+            video_file, vcodec="libx265", crf=custom_bitrate, preset="ultrafast"
+        ).overwrite_output().run(capture_stdout=True, capture_stderr=True)
```
Comment (SpEcHiDe) [inline plugins/ff_mpeg.py:79]: you should use run_in_thread, I think 🤔

Comment (yshalsager) [reply plugins/ff_mpeg.py:79]: Run function is from python ffmpeg package, not python stdlib.

Comment (yshalsager) [reply plugins/ff_mpeg.py:79]: Alright, there are run_async in ffmpeg package, I will check it.

Comment (SpEcHiDe) [reply plugins/ff_mpeg.py:79]: both run and run_async are using subprocess: https://github.com/kkroening/ffmpeg-python/blob/master/ffmpeg/_run.py#L288

Comment (SpEcHiDe) [reply plugins/ff_mpeg.py:79]: I meant https://github.com/UsergeTeam/Userge/blob/alpha/userge/core/ext/pool.py#L30

Comment (yshalsager) [reply plugins/ff_mpeg.py]: Handled it in different way, check the last commit.

**Comment by rking32** [comment]
Hi @yshalsager, Thanks for your hard work and sorry for the delay. Seems like it would be very useful and you have added so many methods. We will check this asap :)

**Comment by Harsh-br0** [comment]
> There's no need to use `ffmpegsave` command anymore. All module commands will work with telegram files, local files, and direct links.

@yshalsager hey there , seems we need that save cmd since we can't have any visual feedbacks like conventional android apps that first visualise the changes and exports them in single attempt , but here we need to do hit-and-trial in some scenarios like setting resolutions , etc so to optimise performance , the cmd may gradually affect it by saving last media locally.
So , Can you please evaluate the scenario and present your views further , thanks

**Comment by yshalsager** [comment]
> > There's no need to use `ffmpegsave` command anymore. All module commands will work with telegram files, local files, and direct links.
> 
> @yshalsager hey there , seems we need that save cmd since we can't have any visual feedbacks like in conventional android apps that first visualise the changes and exports them in single attempt , but here we need to do hit-and-trial in some scenarios like setting resolutions , etc so to optimise performance , the cmd may gradually affect it by saving last media locally. So , Can you please evaluate the scenario and present your views further , thanks

I have replaced ffmpegsave function with get_media_path_and_name that utilizes tg_download and url_download from download module (can be replaced with handle_download BTW), so it still visual progress exactly the same way as download module.

**Comment by Krishna-Singhal** [comment]
> I have replaced ffmpegsave function with get_media_path_and_name that utilizes tg_download and url_download from download module (can be replaced with handle_download BTW), so it still visual progress exactly the same way as download module.

I think you are not getting what he is saying, anyways added -d flag to do same thing as he said

**Review by Krishna-Singhal — APPROVED** [review]

**Review by Krishna-Singhal — APPROVED** [review]

**Pull request merged**